### DISCUSSION TITLE
Link audio and video in FCP XML

### DISCUSSION THREAD

**Discussion by changduckGO** [AcademySoftwareFoundation/OpenTimelineIO#1271] [Q&A]
when create XML file using fcp_xml adapter, Can I link clip of video track and audio track? 
they are "unlinked" when open in Premiere Pro. if possible, How Can I link clip?
Thank you.
- `unlink`
![unlink02](https://user-images.githubusercontent.com/62056652/162911597-7566bb48-8b4e-4e87-930b-0e225977835e.jpg)
![unlink01](https://user-images.githubusercontent.com/62056652/162911604-f28afff7-f921-4855-bed2-28a3f3830625.jpg)

- `link`
![link](https://user-images.githubusercontent.com/62056652/162912047-8018aa52-c644-473e-ad44-236c5aebfff8.jpg)

**Comment by json-lim** [comment]
@changduckGO I'm also trying to figure this one out. By any chance, were you able to find a solution for this issue?

**Comment by json-lim** [comment]
Just to circle back here, I don't think this is possible to achieve at the moment. To link files together, `<link>` metadata tags need to be added to the xml file. However:
1. OpenTimelineIO does not seem to generate link tags at the moment
2. OpenTimelineIO explicitly ignores any `link` input metadata [[github](https://github.com/AcademySoftwareFoundation/OpenTimelineIO/blame/main/src/py-opentimelineio/opentimelineio/adapters/fcp_xml.py#L428)] so I think we are unable to manually add this metadata even if we wanted to.

**Reply by changduckGO** [reply]
I'm sorry to reply too late.

In my case, when creating `otio.schema.Clip`, add directly `@id` in **metadata** as below.

```
metadata={
                     "`fcp_xml`": {
                            "`@id`": _id
                    }
```

Then, add directly `<link>` tag to xml file using `xml.etree.ElementTree`.

**Comment by ssteinbach** [comment]
You could modify the XML adapter to capture this metadata and add a tag to the metadata, and then round trip it.  However, OTIO doesn't provide a way of addressing a reference to another entity, so it would be brittle in the sense that any kind of manipulation of the timeline wouldn't update that reference.  There has been debate around this topic before, but we haven't built anything to support this officially yet, AFAIK. @reinecke might know more about this.

**Comment by meshula** [comment]
OTIO doesn't have a concept of FCP style linking in the manner proposed in this thread, i.e.

```
video track
    clip
      |
     link ----
              |
audio track   |
              |
          -----
     clip
```

but the FCP linking concept can be properly modeled. It would look like this

```
track
    ... stuff preceding the linked clips
    stack
        clip with video media
        track (kind audio)
           gap (to record offsetting information
           clip with audio media
   ... stuff after the linked clips
```

This would correctly model the concept in FCP. In order to enforce the behavioral constraint of linking in FCP for round tripping, the stack's audio track would have to have nothing else in it, just the gap followed by a clip.

OTIO doesn't have any notion of constraints at all, there's no way to tag that special track as being of audio kind, and containing only a gap followed by a clip. However, metadata could be added to the stack to indicate that it represents a FCP style linked pair.

I played with other ways of representing what FCP is doing, and I am not satisfied enough with them to report those constructions here.

I think it is important to properly model the behavior in OTIO, rather than adding heuristic meanings around metadata.

Hopefully these notes spark some ideas.

**Comment by ojadeyemi** [comment]
just stumbled across this issue also. Has there been a solution? How would this be manually implemented if I had too? 

I am using the python package with the fcp_xml otio adapters.
